### ISSUE TITLE
Experiment: Flow Matching (FM)

### ISSUE THREAD

**Issue by csmangum** [Dooders/AgentFarm#98]
Explore the Flow Matching (FM) technique for generative modeling to understand its potential for training Continuous Normalizing Flows (CNFs) without requiring simulation. This experiment aims to implement FM, test it on simple distributions, and evaluate its applicability in a computational model of reality.

---

**Tasks and Steps**

#### **1. Environment Setup**
- [ ] Install required libraries:
  - [ ] Install PyTorch.
  - [ ] Install or configure a library for CNFs, such as `Torchdyn` or `TorchCFM`, if available.
- [ ] Set up the project structure with the following components:
  - **`src/`**: For implementation files.
  - **`notebooks/`**: For experiment visualizations and analysis.
  - **`docs/`**: For documenting findings and observations.
  - **`tests/`**: For validating components of the implementation.

#### **2. Define Source and Target Distributions**
- [ ] Create a function to sample from a source distribution (e.g., standard Gaussian).
- [ ] Create a function to sample from a target distribution:
  - [ ] Start with a simple 2D Gaussian mixture distribution.
  - [ ] Extend to other distributions like high-dimensional Gaussian or discrete data distributions.

#### **3. Implement the Vector Field**
- [ ] Design a neural network to represent the vector field:
  - [ ] Use a simple feedforward network with configurable depth and width.
  - [ ] Include ReLU or other activation functions.
  - [ ] Support input of time-conditioned vectors for learning time-dependent paths.

#### **4. Define the Flow Matching Loss**
- [ ] Implement the FM loss function to train the vector field:
  - [ ] Calculate interpolated points between source and target distributions.
  - [ ] Compute target velocities (difference between source and target points).
  - [ ] Train the network to minimize the MSE between predicted and target velocities.

#### **5. Train the Model**
- [ ] Set up a training loop:
  - [ ] Sample batches from source and target distributions.
  - [ ] Train the vector field neural network using the FM loss.
  - [ ] Monitor and log the loss during training.

#### **6. Experiment with Configurations**
- [ ] Test various neural network architectures:
  - [ ] Adjust the number of layers, neurons, and activation functions.
- [ ] Experiment with different interpolation methods (e.g., linear, nonlinear).
- [ ] Vary the training hyperparameters:
  - Learning rate.
  - Batch size.
  - Number of epochs.

#### **7. Visualize Results**
- [ ] Create plots to visualize:
  - [ ] Source and target distributions.
  - [ ] Interpolated points during training.
  - [ ] Vector field arrows overlaid on the interpolated points.
- [ ] Generate samples from the trained model to validate the quality of the learned transformation.

#### **8. Evaluate Performance**
- [ ] Compare FM against baseline methods like traditional normalizing flows or diffusion models.
- [ ] Analyze:
  - Convergence speed.
  - Quality of the generated samples.
  - Stability during training.

#### **9. Document Findings**
- [ ] Summarize results in a Markdown file or Jupyter Notebook:
  - [ ] Training insights (e.g., convergence trends, loss curves).
  - [ ] Observations on vector field behavior and sample quality.
  - [ ] Challenges and limitations encountered during the experiment.
- [ ] Include visualizations and explanations of the results.

---

**Acceptance Criteria**
- [ ] A complete implementation of Flow Matching for a basic source-to-target mapping task.
- [ ] Multiple visualizations demonstrating the learned transformations.
- [ ] Documentation of the experiment, highlighting key insights and next steps.

---

**Optional Extensions**
- [ ] Test FM on high-dimensional data distributions.
- [ ] Explore applications in image or video generative tasks.
- [ ] Investigate FM's potential for discrete data modeling (e.g., language or sequences).

---

**Additional Resources**
- [ ] [[Flow Matching Paper](https://arxiv.org/)](https://arxiv.org/)

**Comment by csmangum** [comment]
Incorporating Flow Matching (FM) into your computational model of reality and experience could provide a novel way to model and simulate transitions between states in your system. Here’s how you might approach integrating FM principles and techniques into your simulation framework:

---

### 1. **Conceptual Alignment with Your Model**
- **Mapping Source and Target Distributions**: In your computational model of reality, you could interpret the "source distribution" as the initial state of agents, environments, or phenomena, and the "target distribution" as their evolved or desired state. FM's ability to model transitions between distributions aligns well with dynamic processes in your simulation.
- **Conditional Paths as Realistic Transitions**: FM's regression of vector fields for fixed conditional paths allows you to model smooth, realistic transitions in agent states, environmental factors, or interactions without relying on predefined rules or extensive simulations.

---

### 2. **Simulation-Free Transition Modeling**
- **Agent State Transitions**: Instead of simulating detailed step-by-step transitions in agent states, FM could learn the vector field that defines how agents move from one distribution of states to another. For example, you could train FM on transitions between agent behaviors or resource distributions.
- **Environmental Dynamics**: Use FM to model global changes in the environment, such as resource allocation, weather patterns, or population shifts, without explicitly simulating every micro-interaction.

---

### 3. **Optimal Transport Paths for Evolutionary Dynamics**
- **Evolutionary Adaptations**: FM’s alignment with Optimal Transport (OT) principles can model evolutionary or adaptive processes. Agents' genetic parameters (e.g., in your DNA-inspired system) can transition optimally from one generation to the next, minimizing "energy" (e.g., resource use, learning cost).
- **Agent-Environment Feedback Loops**: Incorporate FM to define smooth, optimal transitions in feedback loops where agents adapt to environmental changes, and the environment adapts to agent activity.

---

### 4. **Improved Computational Efficiency**
- **Reduced Complexity in Training**: Since FM does not require forward-backward simulations like diffusion models, it can reduce the computational cost of training your generative models. This efficiency could allow for larger-scale experiments or higher fidelity modeling of agent interactions.
- **Scalability**: Use FM to model complex, high-dimensional distributions in your system, such as multi-agent interactions, without significantly increasing computational requirements.

---

### 5. **Integration with Neural Networks**
- **Neural Network Parameter Evolution**: In your dynamically adaptive evolutionary system, FM can be used to model the interpolation of neural network parameters (weights) during crossover and mutation. This ensures smooth transitions in the networks' functional behavior.
- **Policy Learning**: Agents' choice and action models can be trained using FM to generalize vector fields that represent optimal policies across various scenarios.

---

### 6. **Enhanced Visualization and Analysis**
- **Visualizing State Transitions**: Use FM to generate interpretable, smooth visualizations of how states evolve over time, helping you better understand and analyze your system's dynamics.
- **Exploring "What-If" Scenarios**: FM's ability to generalize paths allows you to explore alternate realities or scenarios by defining new source and target distributions and observing how agents or environments transition.

---

### 7. **Potential Applications in Your Model**
- **Behavioral Modeling**: Use FM to model how agents transition between behavioral states based on internal and external stimuli.
- **Resource Distribution**: Model optimal redistribution of resources using OT-based FM to achieve efficient and realistic outcomes.
- **Conflict and Collaboration**: FM can help simulate smooth transitions between competitive and cooperative dynamics in your system.

---

### 8. **Technical Integration**
- **Frameworks**: Utilize libraries like TorchCFM to implement FM in your system. Train FM models using your existing simulation data as a basis for the source and target distributions.
- **Custom Architectures**: Adapt FM's neural network architectures to align with your system's specific requirements, such as discrete states or continuous parameters.
.

**Comment by csmangum** [comment]
Here's a simple example of implementing Flow Matching (FM) in Python using PyTorch. This example demonstrates training a simple vector field to map a source distribution (e.g., Gaussian noise) to a target distribution (e.g., a 2D Gaussian mixture).

---

### Prerequisites
Make sure you have PyTorch installed:

```bash
pip install torch
```

---

### Code Example

```python
import torch
import torch.nn as nn
import torch.optim as optim
import matplotlib.pyplot as plt

# Define the source distribution (e.g., standard Gaussian)
def sample_source(batch_size):
    return torch.randn(batch_size, 2)

# Define the target distribution (e.g., Gaussian Mixture)
def sample_target(batch_size):
    mixture_weights = torch.tensor([0.5, 0.5])
    means = torch.tensor([[-2, -2], [2, 2]])
    cov = torch.eye(2) * 0.5

    components = torch.distributions.MultivariateNormal(means, cov)
    component_idx = torch.multinomial(mixture_weights, batch_size, replacement=True)
    samples = components.sample()[component_idx]
    return samples

# Neural network for vector field
class VectorField(nn.Module):
    def __init__(self):
        super(VectorField, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(2, 64),
            nn.ReLU(),
            nn.Linear(64, 64),
            nn.ReLU(),
            nn.Linear(64, 2)
        )

    def forward(self, x):
        return self.net(x)

# Flow Matching Loss
def flow_matching_loss(vector_field, source_samples, target_samples, t=0.5):
    # Interpolated samples
    interpolated_samples = (1 - t) * source_samples + t * target_samples
    
    # Target velocities
    target_velocities = target_samples - source_samples
    
    # Predicted velocities from the vector field
    predicted_velocities = vector_field(interpolated_samples)
    
    # MSE loss
    loss = nn.MSELoss()(predicted_velocities, target_velocities)
    return loss

# Training
def train_flow_matching(vector_field, num_epochs=1000, batch_size=256, lr=1e-3):
    optimizer = optim.Adam(vector_field.parameters(), lr=lr)

    for epoch in range(num_epochs):
        # Sample source and target distributions
        source_samples = sample_source(batch_size)
        target_samples = sample_target(batch_size)

        # Compute loss and backpropagate
        loss = flow_matching_loss(vector_field, source_samples, target_samples)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if (epoch + 1) % 100 == 0:
            print(f"Epoch {epoch+1}/{num_epochs}, Loss: {loss.item()}")

# Visualization
def visualize_mapping(vector_field):
    with torch.no_grad():
        # Sample source and target distributions
        source_samples = sample_source(500).numpy()
        target_samples = sample_target(500).numpy()

        # Transform source samples using the learned vector field
        t = 0.5  # Midpoint
        interpolated_samples = (1 - t) * source_samples + t * target_samples
        vector_field_outputs = vector_field(torch.tensor(interpolated_samples).float()).numpy()

        # Plot
        plt.figure(figsize=(8, 6))
        plt.scatter(source_samples[:, 0], source_samples[:, 1], alpha=0.6, label='Source Samples')
        plt.scatter(target_samples[:, 0], target_samples[:, 1], alpha=0.6, label='Target Samples')
        plt.quiver(
            interpolated_samples[:, 0], interpolated_samples[:, 1],
            vector_field_outputs[:, 0], vector_field_outputs[:, 1],
            angles='xy', scale_units='xy', scale=1, color='r', alpha=0.7, label='Learned Vector Field'
        )
        plt.legend()
        plt.title("Flow Matching: Source to Target")
        plt.xlabel("x1")
        plt.ylabel("x2")
        plt.show()

# Main
if __name__ == "__main__":
    # Initialize vector field
    vector_field = VectorField()

    # Train vector field
    train_flow_matching(vector_field)

    # Visualize results
    visualize_mapping(vector_field)
```

---

### Explanation
1. **Source and Target Distributions**:  
   - The source is a standard Gaussian.
   - The target is a Gaussian Mixture Model with two components.

2. **Vector Field**:  
   A neural network learns the vector field that moves points from the source distribution to the target distribution.

3. **Loss Function**:  
   The Flow Matching loss minimizes the mean squared error (MSE) between the predicted velocities from the vector field and the actual velocities (target - source).

4. **Training**:  
   The vector field is trained using samples from the source and target distributions.

5. **Visualization**:  
   A quiver plot shows the learned vector field moving the source samples toward the target distribution.

---

### Output
- The quiver plot will display arrows (the vector field) illustrating the learned transformation from the source to the target distribution.

**Continuous Normalizing Flows (CNFs)** are a type of generative model that learns to map one probability distribution into another using continuous transformations. They are a special class of normalizing flows that use ordinary differential equations (ODEs) to describe the transformation process. Here's a breakdown of the concept:

---

### Key Components of CNFs:

1. **Normalizing Flows**:
   - Normalizing flows are a family of methods that transform a simple (base) probability distribution (e.g., Gaussian) into a more complex target distribution by applying a series of invertible transformations.
   - Each transformation's Jacobian determinant can be computed to track changes in the probability density.

2. **Continuous Transformations**:
   - Unlike standard normalizing flows that use discrete transformations, CNFs use a continuous transformation defined by an ODE:
     \[
     \frac{dz(t)}{dt} = f(z(t), t; \theta)
     \]
     - \(z(t)\): Data at time \(t\).
     - \(f(z, t; \theta)\): A neural network parameterized by \(\theta\) that defines the dynamics of the transformation.
   - This equation evolves \(z(t)\) from an initial distribution (\(z(0)\)) to a target distribution (\(z(1)\)).

3. **Probability Density Dynamics**:
   - The probability density function (PDF) of the transformed distribution can be computed using the **change of variables formula** derived from the ODE:
     \[
     \log p(z(t)) = \log p(z(0)) - \int_0^t \text{tr}\left(\frac{\partial f}{\partial z}\right) \, dt
     \]
     - The trace of the Jacobian of \(f\) captures how the volume changes during the transformation.

4. **Training Objective**:
   - CNFs are trained to maximize the likelihood of the target data by optimizing the parameters \(\theta\) of the neural network \(f(z, t; \theta)\).
   - This is achieved by minimizing the difference between the transformed source distribution and the target data distribution.

---

### Why Use CNFs?

1. **Flexibility**:
   - CNFs can model complex target distributions with fewer parameters compared to discrete normalizing flows.

2. **Invertibility**:
   - CNFs are inherently invertible, allowing efficient computation of both the forward (sampling) and inverse (density evaluation) processes.

3. **Scalability**:
   - CNFs can handle high-dimensional data more efficiently by leveraging continuous transformations.

---

### Applications of CNFs:

1. **Generative Modeling**:
   - CNFs are used to generate samples from complex distributions (e.g., images, audio, video).
   - Example: Modeling ImageNet distributions.

2. **Density Estimation**:
   - CNFs provide exact likelihood estimates, making them suitable for probabilistic modeling.

3. **Physics-Inspired Models**:
   - CNFs can represent physical systems where dynamics are naturally described by differential equations (e.g., fluid dynamics, molecular simulations).

4. **Optimal Transport**:
   - CNFs can be used to compute optimal transport maps between probability distributions.

---

### Example of CNF Workflow:

1. Define a simple base distribution (e.g., Gaussian).
2. Train a neural network \(f(z, t; \theta)\) to transform the base distribution into a complex target distribution by solving the ODE.
3. Use the trained model to:
   - Sample new data points by evolving the base distribution through the learned dynamics.
   - Compute likelihoods of data points.

---

### Libraries for CNFs:
- **Torchdyn**: A PyTorch-based library for working with neural ODEs and CNFs.
- **DiffEqFlux.jl**: A Julia library for solving differential equations and implementing CNFs.
- **FFJORD**: An implementation of CNFs in PyTorch.

---

In summary, CNFs are a powerful technique for generative modeling and density estimation, leveraging continuous transformations described by differential equations to learn flexible mappings between distributions.
